tests/Tests/ORM/Functional/ValueConversionType/OneToManyCriteriaMatchingTest.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Functional\ValueConversionType;
+
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\Common\Collections\Expr\Comparison;
+use Doctrine\Tests\Models\ValueConversionType\InversedOneToManyEntity;
+use Doctrine\Tests\Models\ValueConversionType\OwningManyToOneEntity;
+use Doctrine\Tests\OrmFunctionalTestCase;
+use Generator;
+
+class OneToManyCriteriaMatchingTest extends OrmFunctionalTestCase
+{
+    protected function setUp(): void
+    {
+        $this->useModelSet('vct_onetomany');
+
+        parent::setUp();
+    }
+
+    /**
+     * @dataProvider provideMatchingExpressions
+     */
+    public function testCriteriaMatchingOnFieldInOneToManyTarget(Comparison $comparison): void
+    {
+        $entityWithCollection      = new InversedOneToManyEntity();
+        $entityWithCollection->id1 = 'associated';
+
+        $matching                   = new OwningManyToOneEntity();
+        $matching->id2              = 'first';
+        $matching->field            = 'match this';
+        $matching->associatedEntity = $entityWithCollection;
+
+        $notMatching                   = new OwningManyToOneEntity();
+        $notMatching->id2              = 'second';
+        $notMatching->field            = 'this is no match';
+        $notMatching->associatedEntity = $entityWithCollection;
+
+        $this->_em->persist($entityWithCollection);
+        $this->_em->persist($matching);
+        $this->_em->persist($notMatching);
+        $this->_em->flush();
+        $this->_em->clear();
+
+        $this->getQueryLog()->reset()->enable();
+
+        $entityWithCollection = $this->_em->find(InversedOneToManyEntity::class, 'associated');
+        self::assertFalse($entityWithCollection->associatedEntities->isInitialized(), 'Pre-condition: lazy collection');
+
+        $result = $entityWithCollection->associatedEntities->matching(Criteria::create()->where($comparison));
+
+        $l = $this->getQueryLog();
+
+        self::assertCount(1, $result);
+        self::assertSame('first', $result[0]->id2);
+    }
+
+    public function provideMatchingExpressions(): Generator
+    {
+        yield [Criteria::expr()->eq('field', 'match this')];
+        yield [Criteria::expr()->in('field', ['match this'])];
```

Comment:
Not sure if it's related to custom types at all, but this test fails because this code doesn't properly handle the `IN` operator: https://github.com/doctrine/orm/blob/fd041fbe80ee5a743bc45bb1abfe3a4b5d6d29e2/src/Persisters/Collection/ManyToManyPersister.php#L257-L263

It should:
1. Add parenthesis to the SQL.
2. Repeat the placeholder, `$params[]` and `$paramTypes[]` for each element of the value.